sql/expression/function/aggregation/window_framer.go

```diff
@@ -397,13 +467,83 @@ func (f *rangeFramerBase) NewFramer(interval sql.WindowInterval) sql.WindowFrame
 		startNFollowing:    f.startNFollowing,
 		endNPreceding:      f.endNPreceding,
 		endNFollowing:      f.endNFollowing,
+		// range specific
+		startInclusion: startInclusion,
+		endInclusion:   endInclusion,
+	}, nil
+}
+
+func (f *rangeFramerBase) Next(ctx *sql.Context, buf sql.WindowBuffer) (sql.WindowInterval, error) {
+	if f.idx != 0 && f.idx >= f.partitionEnd || !f.partitionSet {
+		return sql.WindowInterval{}, io.EOF
+	}
+
+	var err error
+	newStart := f.frameStart
+	switch {
+	case newStart < f.partitionStart, f.unboundedPreceding:
+		newStart = f.partitionStart
+	default:
+		newStart, err = findInclusionBoundary(ctx, f.idx, newStart, f.partitionEnd, f.startInclusion, f.expr, buf, greaterThanOrEqual)
+		if err != nil {
+			return sql.WindowInterval{}, nil
+		}
+	}
+
+	newEnd := f.frameEnd
+	if newStart > newEnd {
+		newEnd = newStart
+	}
+	switch {
+	case newEnd > f.partitionEnd, f.unboundedFollowing:
+		newEnd = f.partitionEnd
+	default:
+		newEnd, err = findInclusionBoundary(ctx, f.idx, newEnd, f.partitionEnd, f.endInclusion, f.expr, buf, greaterThan)
+		if err != nil {
+			return sql.WindowInterval{}, nil
+		}
 	}
+
+	f.idx++
+	f.frameStart = newStart
+	f.frameEnd = newEnd
+	return f.Interval()
 }
 
-func (f *rangeFramerBase) Next() (sql.WindowInterval, error) {
-	// TODO pass order by clause in constructor
-	// TODO scan ahead buffer of peer groups, use rank algorithm
-	panic("implement me")
+type stopCond int
+
+const (
+	greaterThan        = 1
+	greaterThanOrEqual = 0
+)
+
+// findInclusionBoundary searches a sorted [buffer] for the last index satisfying
+// the comparison: [inclusion] [stopCond] [expr]. For example, (x+2) > (x). This
+// is used as a sliding window algorithm for value ranges.
+func findInclusionBoundary(ctx *sql.Context, pos, searchStart, partitionEnd int, inclusion, expr sql.Expression, buf sql.WindowBuffer, stopCond stopCond) (int, error) {
+	cur, err := inclusion.Eval(ctx, buf[pos])
+	if err != nil {
+		return 0, err
+	}
+
+	i := searchStart
+	for i < partitionEnd {
+		res, err := expr.Eval(ctx, buf[i])
+		if err != nil {
+			return 0, err
+		}
+
+		cmp, err := expr.Type().Compare(res, cur)
+		if err != nil {
+			return 0, err
+		}
+
+		if cmp >= int(stopCond) {
+			break
+		}
+		i++
```

Comment:
Typically better to write this as a traditional for loop in this case

Reply:
You can use an empty statement for the initialization since you want to declare `i` outside the loop